### ISSUE TITLE
Implement AST

### ISSUE THREAD

**Issue by ChangeCaps** [cjavad/dare#14]
Before we can run the core algorithm we must parse our `TokenStream` into a tree form (abstract syntax tree). 

This issue is just about the AST itself not the parser as I think that should be implemented and reasoned about in a separate issue.

I suggest implementing it something along the lines of this:
```rust
enum UnaryOperatorKind;

struct UnaryOperator {
    kind: UnaryOperatorKind,
    span: Span,
}

enum BinaryOperatorKind;

struct BinaryOperator {
    kind: BinaryOperatorKind,
    span: Span,
}

struct AtomicExpression {
    ident: String,
}

struct UnaryExpression {
    operator: UnaryOperator,
    operand: Expression,
}

struct BinaryExpression {
    lhs: Expression,
    operator: BinaryOperator,
    rhs: Expression,
}

struct ParenExpression {
    expression: Expression,
}

enum ExpressionKind {
    Atomic(AtomicExpression),
    Paren(ParenExpression),
    Unary(UnaryExpression),
    Binary(BinaryExpression),
}

struct Expression {
    kind: ExpressionKind,
    span: Span,
}
```

This way is a span at every node of the AST, allowing us to write better errors.
